src/Symfony/Component/VarDumper/Caster/ReflectionCaster.php

```diff
@@ -96,11 +96,21 @@ public static function castType(\ReflectionType $c, array $a, Stub $stub, $isNes
     {
         $prefix = Caster::PREFIX_VIRTUAL;
 
-        $a += [
-            $prefix.'name' => $c instanceof \ReflectionNamedType ? $c->getName() : (string) $c,
-            $prefix.'allowsNull' => $c->allowsNull(),
-            $prefix.'isBuiltin' => $c->isBuiltin(),
-        ];
+        if ($c instanceof \ReflectionUnionType) {
+            $a[$prefix.'allowsNull'] = $c->allowsNull();
+            self::addMap($a, $c, [
+                'types' => 'getTypes',
+            ]);
+        } elseif ($c instanceof \ReflectionNamedType) {
+            $a += [
+                $prefix.'name' => $c->getName(),
+                $prefix.'allowsNull' => $c->allowsNull(),
+                $prefix.'isBuiltin' => $c->isBuiltin(),
+            ];
+        } else {
+            // As of PHP 8.0 there are only two children of ReflectionType (abstract) so this is unreachable.
+            $a[$prefix.'allowsNull'] = $c->allowsNull();
```

Comment:
should we rather throw an `UnexpectedValueException` in this currently unreadable branch, so that we get a proper reporting if it becomes reachable ?

Reply:
another option (potentially cleaner) is to split casters for ReflectionUnionType and ReflectionNamedType

Reply:
> should we rather throw an UnexpectedValueException in this currently unreadable branch, so that we get a proper reporting if it becomes reachable ?

I'm happy to update with the exception, if that is preferable.  I'm not sure what the Symfony change management process is like so I'm happy to go with your suggestion here if you think that is a better fit.

> another option (potentially cleaner) is to split casters for ReflectionUnionType and ReflectionNamedType

The logic for handling the abstract ReflectionType is minimal, if we split it into two functions then everywhere someone calls this function they would need to check type before calling.  If the return types were significantly different that could be preferable, but this is building a fairly complex array, almost like a view, so I'm unclear on the benefit. 

We could possibly then create a third function that wraps the other two and does the type checking proxy but, to me, that seems like it would be harder to read if the differences remain small.

Reply:
> The logic for handling the abstract ReflectionType is minimal, if we split it into two functions then everywhere someone calls this function they would need to check type before calling.

That change would be too heavy for a bugfix, I think. I'm fine with the `if`, but I agree that we should throw if the unreachable `else` is actually hit.

Reply:
Apologies, I know my comment here led us down this particular path, but I was just thinking about how I would test this and I realized it is actually possible.

```
class FakeReflection extends \ReflectionType {
    public function allowsNull() {
        return false;
    }

    public function __toString()
    {
       return 'fake';
    }
};
$t = new FakeReflection();
echo match(true) {
    $t instanceof \ReflectionNamedType => 'named',
    $t instanceof \ReflectionUnionType => 'union',
    $t instanceof \ReflectionType => 'abstract',
    default => 'error'
};
```
```> 'abstract'```

Reply:
Maybe just remove the comment?

Reply:
I'm going to mark this resolved as it was based on a misconception that this is unreachable.